src/azul/plugins/metadata/hca/transform.py

```diff
@@ -940,6 +943,27 @@ def _sequencing_input(self, sequencing_input: api.Biomaterial) -> MutableJSON:
             'sequencing_input_type': api.schema_names[type(sequencing_input)]
         }
 
+    @classmethod
+    def _aggregate_date_types(cls) -> FieldTypes:
+        return {
+            'document_id': null_str,
+            'submission_date': null_datetime,
+            'update_date': null_datetime,
+        }
+
+    def _aggregate_date(self) -> MutableJSON:
+        submission_date = None
+        update_date = None
+        for entity in api.not_stitched(self.api_bundle.entities.values()):
+            submission_date = min((submission_date, entity.submission_date),
+                                  key=none_safe_key(none_last=True))
+            update_date = max((update_date, entity.update_date), key=none_safe_key())
```

Comment:
This is sort of the worst of both paradigm, the functional and the imperative. It creates identical pairs of callables for each iteration and each callable creates two tuples when called. So that's n * 6 objects being created. How about this:

```diff
Index: src/azul/plugins/metadata/hca/transform.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/metadata/hca/transform.py b/src/azul/plugins/metadata/hca/transform.py
--- a/src/azul/plugins/metadata/hca/transform.py	(revision d9ca31f9f6c93e7e60cc7471a0b62b1448e1f026)
+++ b/src/azul/plugins/metadata/hca/transform.py	(date 1631575124983)
@@ -952,16 +952,13 @@
         }
 
     def _aggregate_date(self) -> MutableJSON:
-        submission_date = None
-        update_date = None
-        for entity in api.not_stitched(self.api_bundle.entities.values()):
-            submission_date = min((submission_date, entity.submission_date),
-                                  key=none_safe_key(none_last=True))
-            update_date = max((update_date, entity.update_date), key=none_safe_key())
+        entities = api.not_stitched(self.api_bundle.entities.values())
+        submission_dates = (e.submission_date for e in entities)
+        update_dates = (e.update_date for e in entities)
         return {
             'document_id': str(self.api_bundle.uuid),
-            'submission_date': submission_date,
-            'update_date': update_date,
+            'submission_date': min(filter(None, submission_dates), default=None),
+            'update_date': max(filter(None, update_dates), default=None)
         }
 
     @classmethod

```